contributing/specifications/functions.md

```diff
@@ -0,0 +1,111 @@
+# Composition Functions Specification
+
+A Composition Function is a kind of Crossplane extension. A Function is called
+to instruct Crossplane how a Composite Resource (XR) should be composed of other
+resources. This document specifies how a Function MUST behave. Refer to the
+Composition Functions [design document] for broader context on the feature.
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are
+to be interpreted as described in [RFC 2119].
+
+## Serving Requests from Crossplane
+
+A Function MUST implement a gRPC `FunctionRunnerService` server. A Function
+SHOULD implement the latest available version of this service - e.g. `v1beta1`.
+The authoritative definition of this service can be found at the following URL.
+
+https://github.com/crossplane/crossplane/tree/master/apis/apiextensions/fn/proto
+
+A Function MUST copy the tag field from a RunFunctionRequest's RequestMeta
+message to the ResponseMeta tag field of the corresponding RunFunctionResponse.
+A Function MUST NOT modify the tag field. A Function MUST NOT inspect or
+otherwise depend on the tag's value
+
+A Function SHOULD specify the ttl field of each RunFunctionResponse. The value
+of the ttl field MUST reflect how long Crossplane may use a cached version of
+the response for an identical RunFunctionRequest.
+
+A Function MUST pass through any desired state for which it does not have an
+opinion from the RunFunctionRequest desired field to the corresponding
+RunFunctionResponse desired field. A Function MAY mutate any desired state for
+which it has an opinion, including adding to, updating, or deleting from the
+desired state. For example a Function may:
+
+* Add a new composed resource to the desired state.
+* Add a new field to an existing composed resource in the desired state.
+* Remove a composed resource from the desired state.
+* etc...
+
+A Function MUST do so _intentionally_. Put otherwise, remember that Functions
+are run in a pipeline. A Function MUST NOT accidentally forget to propagate the
+desired state passed to it by previous Functions.
+
+A Function MUST NOT specify desired composite resource spec or metadata fields.
+A Function MAY only specify desired composite resource status fields.
+
+A Function MUST NOT specify desired composed resource status fields. A Function
+MAY only specify non-status desired composed resource fields, for example fields
+under the top-level metadata, spec, or data fields.
+
+A Function SHOULD NOT specify a metadata.name for desired composed resources.
```

Comment:
SHOULD NOT or MUST NOT? This reads like a `MUST NOT`

Reply:
This one's actually SHOULD NOT - you can but it's not recommended.

Reply:
Is it worth adding something like "if a `metadata.name` is specified it MUST be unique for the associated GVK within the kubernetes cluster"?   I have had several questions in Slack this week from people who want to set their own `metadata.name` values for XRs and MRs.